staging/src/k8s.io/apimachinery/pkg/util/cache/expiring.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"container/heap"
+	"context"
+	"sync"
+	"time"
+
+	utilclock "k8s.io/apimachinery/pkg/util/clock"
+)
+
+// NewExpiring returns an initialized expiring cache that's ready to use. The
+// GC thread will shutdown when the passed context is closed.
+func NewExpiring(ctx context.Context) *Expiring {
+	return NewExpiringWithClock(ctx, utilclock.RealClock{})
+}
+
+// NewExpiringWithClock is like NewExpiring but allows passing in a custom
+// clock for testing.
+func NewExpiringWithClock(ctx context.Context, clock utilclock.Clock) *Expiring {
+	e := &Expiring{
+		clock: clock,
+		m:     make(map[interface{}]entry),
+	}
+	go e.start(ctx)
+	return e
+}
+
+// Expiring is a map whose entries expire after a per-entry timeout.
+type Expiring struct {
+	clock utilclock.Clock
+
+	// mu protects the below fields
+	mu sync.RWMutex
+	// m is the internal map that backs the cache.
+	m map[interface{}]entry
+	// generation is a generation number so that when setting multiple entries
+	// with the same key, the cleanup for earlier versions of the entry won't
+	// remove later ones.
+	generation uint64
+
+	heap expiringHeap
+}
+
+type entry struct {
+	val        interface{}
+	expiry     time.Time
+	generation uint64
+}
+
+// Get looks up an entry in the cache.
+func (c *Expiring) Get(key interface{}) (val interface{}, ok bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	e, ok := c.m[key]
+	if !ok || !c.clock.Now().Before(e.expiry) {
+		return nil, false
+	}
+	return e.val, true
+}
+
+// Set sets a key/value/expiry entry in the map, overwriting any previous entry
+// with the same key. The entry expires at the given expiry time, but it may be
+// removed earlier by other Set calls.
+func (c *Expiring) Set(key interface{}, val interface{}, ttl time.Duration) {
+	expiry := c.clock.Now().Add(ttl)
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.generation++
```

Comment:
E.g.

```go
func (c *Expiring) nextGenerationLocked() int64 {
  g := c.generation
  c.generation++
  return g
}
```

In this case it's optional since a) you're actually tracking the last value given rather than the next value to be used (which I hadn't realized when I wrote that comment) and b) this is the only place that needs to consume a generation.